src/TiledArray/initializer_list_utils.h

```diff
@@ -0,0 +1,313 @@
+    out_itr = std::copy(il.begin(), il.end(), out_itr);
+  }
+  // The initializer list is at least a matrix, so recurse over sub-lists
+  else{
+    const auto length = il.begin()->size();
+    for(auto&& x : il){
+      TA_ASSERT(x.size() == length); // sub-lists must be the same size
+      out_itr = flatten_il(x, out_itr);
+    }
+  }
+  return out_itr;
+}
+
+//------------------------------------------------------------------------------
+// array_from_il free function
+//------------------------------------------------------------------------------
+
+/** @brief Converts an `std::initializer_list` into an array.
+ *
+ *  This function encapsulates the process of turning an `std::initializer_list`
+ *  into a TiledArray array. The resulting tensor will consistent of a single
+ *  tile which holds all of the values.
+ *
+ *  @note This function will raise a static assertion if you try to construct a
+ *        rank 0 tensor (*i.e.*, you pass in a single element and not an
+ *        `std::initializer_list`).
+ *
+ * @tparam ArrayType The type of the array we are creating. Expected to be
+ *                   have an API akin to that of DistArray
+ * @tparam T The type of the provided `std::initializer_list`.
+ *
+ * @param[in] world The context in which the resulting tensor will live.
+ * @param[in] il The initializer_list containing the initial state of the
+ *               tensor. @p il is assumed to be non-empty and in row-major form.
+ *               The nesting of @p il will be used to determine the rank of the
+ *               resulting tensor.
+ *
+ * @return A newly created instance of type @p ArrayType whose state is derived
+ *         from @p il and exists in the @p world context.
+ *
+ * @throw TiledArray::Exception if @p il contains no elements. If an exception
+ *                              is raised @p world and @p il are unchanged
+ *                              (strong throw guarantee).
+ * @throw TiledArray::Exception If the provided `std::initializer_list` is not
+ *                              rectangular (*e.g.*, attempting to initialize
+ *                              a matrix with the value `{{1, 2}, {3, 4, 5}}`).
+ *                              If an exception is raised @p world and @p il are
+ *                              unchanged.
+ */
+template<typename ArrayType, typename T>
+auto array_from_il(World& world, T&& il) {
+  using element_type = typename ArrayType::element_type;
+  using tile_type    = typename ArrayType::value_type;
+
+  static_assert(initializer_list_rank_v<std::decay_t<T>> > 0,
+      "value initializing rank 0 tensors is not supported");
+
+  auto trange = tiled_range_from_il(il);
+
+  // Use the fact that there's only one tile
+  auto range = trange.make_tile_range(0);
+  std::vector<element_type> buffer;
+  buffer.reserve(range.volume());
+  flatten_il(il, buffer.begin());
+
+  ArrayType rv(world, trange);
+  (*rv.begin()) = tile_type(range, buffer.data());
```

Comment:
this will not work if work has multiple ranks since tiles can only be set once. It's best to determine if the lone tile is local to this rank and then assign it from there